src/WebJobs.Script/Binding/Http/HttpTriggerAttributeBindingProvider.cs

```diff
@@ -51,37 +52,57 @@ public Task<ITriggerBinding> TryCreateAsync(TriggerBindingProviderContext contex
                     "Can't bind HttpTriggerAttribute to type '{0}'.", parameter.ParameterType));
             }
 
-            return Task.FromResult<ITriggerBinding>(new HttpTriggerBinding(context.Parameter, isUserTypeBinding));
+            return Task.FromResult<ITriggerBinding>(new HttpTriggerBinding(attribute, context.Parameter, isUserTypeBinding));
         }
 
         internal class HttpTriggerBinding : ITriggerBinding
         {
             private readonly ParameterInfo _parameter;
             private readonly IBindingDataProvider _bindingDataProvider;
             private readonly bool _isUserTypeBinding;
+            private readonly Dictionary<string, Type> _bindingDataContract;
+            private readonly HttpRouteFactory _httpRouteFactory = new HttpRouteFactory("api");
 
-            public HttpTriggerBinding(ParameterInfo parameter, bool isUserTypeBinding)
+            public HttpTriggerBinding(HttpTriggerAttribute attribute, ParameterInfo parameter, bool isUserTypeBinding)
             {
                 _parameter = parameter;
                 _isUserTypeBinding = isUserTypeBinding;
 
+                Dictionary<string, Type> aggregateDataContract = new Dictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
                 if (_isUserTypeBinding)
                 {
                     // Create the BindingDataProvider from the user Type. The BindingDataProvider
                     // is used to define the binding parameters that the binding exposes to other
                     // bindings (i.e. the properties of the POCO can be bound to by other bindings).
                     // It is also used to extract the binding data from an instance of the Type.
                     _bindingDataProvider = BindingDataProvider.FromType(parameter.ParameterType);
+                    if (_bindingDataProvider.Contract != null)
+                    {
+                        foreach (var pair in _bindingDataProvider.Contract)
+                        {
+                            aggregateDataContract.Add(pair.Key, pair.Value);
+                        }
+                    }
                 }
+
+                // add any route parameters to the contract
+                if (!string.IsNullOrEmpty(attribute.RouteTemplate))
+                {
+                    var routeParameters = _httpRouteFactory.GetRouteParameters(attribute.RouteTemplate);
+                    foreach (string parameterName in routeParameters)
+                    {
+                        aggregateDataContract[parameterName] = typeof(string);
+                    }
```

Comment:
Good question. I've put a guard here so we only add if not already present. That way, we retain the actual Type defined in a poco. PropertyHelper SetValue does conversions. I'm adding a test for this.